torch/_inductor/ir.py

```diff
@@ -5567,3 +5578,117 @@ def codegen_collective(self, wrapper, output_name, input_names):
             f"group={output_name}_pg, "
             "async_op=True)"
         )
+
+
+class AllToAllSingle(OutOfPlaceCollectiveKernel):
+    """
+    NOTE: Difference of AllToAllSingle compared to other OutOfPlaceCollectiveKernel:
+
+    1. `create_output_buffers` is not called, because we rely on the underlying
+    `fun_col_impl._all_to_all_single` kernel to do the output buffer allocation.
+    2. `fun_col_impl._register_tensor_work` is not emitted, because the underlying
+    `fun_col_impl._all_to_all_single` kernel already calls it.
+    3. `c10d._find_or_create_pg_by_ranks_and_tag` is not emitted, because the underlying
+    `fun_col_impl._all_to_all_single` kernel already calls it.
+    """
+    def __init__(
+        self,
+        layout,
+        inputs,
+        constant_args,
+        arg_index_to_input_index,
+    ):
+        super().__init__(layout, inputs, [], constant_args)
+        self.arg_index_to_input_index = arg_index_to_input_index
+
+    def should_emit_register_tensor_work(self):
+        return False
+
+    def should_emit_find_or_create_pg(self):
+        return False
+
+    def codegen_output(self, wrapper, output_name, input_names):
+        input_names = [t.codegen_reference() for t in self.original_inputs]
+        wrapper.writeline(f"{output_name}_inputs = [{','.join(input_names)}]")
+
+    @classmethod
+    def create(
+        cls,
+        x: "TensorBox",
+        output_split_sizes: Optional["TensorBox"],
+        input_split_sizes: Optional["TensorBox"],
+        tag: str,
+        ranks: List[int],
+        group_size: int,
+    ):
+        x_realized = cls.realize_input(x)
+        inputs_allow_none = []
+        output_split_sizes_realized = None
+        if output_split_sizes is not None:
+            output_split_sizes_realized = cls.realize_input(output_split_sizes)
+        input_split_sizes_realized = None
+        if input_split_sizes is not None:
+            input_split_sizes_realized = cls.realize_input(input_split_sizes)
+        inputs_allow_none = [
+            x_realized,
+            output_split_sizes_realized,
+            input_split_sizes_realized,
+        ]
+        inputs = [inp for inp in inputs_allow_none if inp is not None]
+        arg_index_to_input_index = {}
+        for i in range(len(inputs_allow_none)):
+            if inputs_allow_none[i] is not None:
+                arg_index_to_input_index[i] = inputs.index(inputs_allow_none[i])
+            else:
+                arg_index_to_input_index[i] = None
+
+        new_size = x_realized.get_size()
+        output_shape_first_dim_s = None
+        if output_split_sizes is not None:
+            output_shape_first_dim_s = V.graph.current_node.meta['val'].size()[0].node.expr
```

Comment:
I feel it may be possible for this size to be an int; e.g., if a user asserts the unbacked symint == 5, then we should have refined it by this point.